src/main/scala/com/microsoft/hyperspace/index/IndexLogEntry.scala

```diff
@@ -111,6 +102,43 @@ object Content {
       None
     }
   }
+
+  /**
+   * Apply `func` to each file in directory recursively.
+   *
+   * @param prefixPath Root prefix
+   * @param directory Root directory
+   * @param func Function which would apply to current prefix and file
+   * @tparam T
+   * @return Result list of applying function to all files
+   */
+  def recFilesApply[T](
+                        prefixPath: Path,
+                        directory: Directory,
+                        func: (FileInfo, Path) => T): Seq[T] = {
+    @tailrec
+    def recAcc[A](
+                   dirMap: List[(Path, Seq[Directory])],
+                   func: (FileInfo, Path) => A,
+                   acc: Seq[A] = Seq.empty): Seq[A] = {
+      dirMap match {
+        case Nil => acc
+        case (curPrefixPath, curDirs) :: otherDirs =>
+          val curAcc = for {
+            dir <- curDirs
+            file <- dir.files
+          } yield func(file, new Path(curPrefixPath, dir.name))
+
+          val newLevels = curDirs
+            .filter(_.subDirs.nonEmpty)
+            .map(dir => (new Path(curPrefixPath, dir.name), dir.subDirs))
+
+          recAcc(otherDirs ++ newLevels, func, curAcc ++ acc)
+      }
+    }
+
+    recAcc(List(prefixPath -> Seq(directory)), func)
```

Comment:
nit: for consistency? 
```suggestion
    recAcc(List((prefixPath, Seq(directory))), func)
```